src/auto-evo/steps/MigrateSpecies.cs

```diff
@@ -0,0 +1,71 @@
+﻿namespace AutoEvo;
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+/// <summary>
+///   Step that generates species migrations for each patch
+/// </summary>
+public class MigrateSpecies : IRunStep
+{
+    private Patch patch;
+    private SimulationCache cache;
+    private Random random;
+
+    public MigrateSpecies(Patch patch, SimulationCache cache)
+    {
+        this.patch = patch;
+        this.cache = cache;
+        random = new Random();
+    }
+
+    public int TotalSteps => 1;
+
+    public bool CanRunConcurrently => true;
+
+    public bool RunStep(RunResults results)
+    {
+        var miche = results.GetMicheForPatch(patch);
+
+        var occupantsSet = new HashSet<Species>();
+        miche.GetOccupants(occupantsSet);
+
+        var occupants = occupantsSet.ToList();
+
+        if (occupants.Count == 0)
+            return true;
+
+        for (int i = 0; i < Constants.AUTO_EVO_MOVE_ATTEMPTS; ++i)
```

Comment:
I think as there's this loop, it would be pretty easy to split this step up into sub-steps to better have the auto-evo progress flow smoothly (which it doesn't do when there's only like 80 steps even at generation 15).